pxd.c

```diff
@@ -1212,24 +1206,59 @@ static int pxd_init_disk(struct pxd_device *pxd_dev)
 	  pxd_dev->tag_set.cmd_size = sizeof(struct fuse_req);
 
 	  err = blk_mq_alloc_tag_set(&pxd_dev->tag_set);
-	  if (err)
-		goto out_disk;
+	  if (err) {
+		return err;
+	  }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,15,0)
+	  disk = blk_mq_alloc_disk(&pxd_dev->tag_set, pxd_dev);
+	  if (IS_ERR(disk)) {
+		blk_mq_free_tag_set(&pxd_dev->tag_set);
+		return PTR_ERR(disk);
+	  }
+	  q = disk->queue;
+#else
+	  /* Create gendisk info. */
+	  disk = alloc_disk(1);
+	  if (!disk) {
+		return -ENOMEM;
+	  }
 	  q = blk_mq_init_queue(&pxd_dev->tag_set);
 	  if (IS_ERR(q)) {
-		err = PTR_ERR(q);
 		blk_mq_free_tag_set(&pxd_dev->tag_set);
-		goto out_disk;
+		put_disk(disk);
+		return PTR_ERR(q);
+	  }
+#endif
+#else /* !__PX_BLKMQ__		  */
+	  /* Create gendisk info. */
+	  disk = alloc_disk(1);
+	  if (!disk) {
+		return -ENOMEM;
 	  }
-#else
 	  q = blk_init_queue(pxd_rq_fn, &pxd_dev->qlock);
 	  if (!q) {
-		err = -ENOMEM;
-	  	goto out_disk;
+		put_disk(disk);
+		return -ENOMEM;
 	  }
 #endif /* __PX_BLKMQ__ */
 #endif /* __PXD_BIO_MAKEREQ__ */
 
+	// Disk and queue initialization
+	snprintf(disk->disk_name, sizeof(disk->disk_name),
+		 PXD_DEV"%llu", pxd_dev->dev_id);
+	disk->major = pxd_dev->major;
+	disk->minors = 1;
+	disk->first_minor = pxd_dev->minor;
+	disk->flags |= GENHD_FL_EXT_DEVT | GENHD_FL_NO_PART_SCAN;
+#ifdef __PXD_BIO_MAKEREQ__
+		disk->fops = get_bd_fpops();
+#else
+		disk->fops = &pxd_bd_ops;
+#endif
```

Comment:
The above will fail compilation in some releases where block_device_operations does not have submit_bio.
Instead only do:
disk->fops = get_bd_fpops() always.

Reply:
done